src/state.rs

```diff
@@ -0,0 +1,219 @@
+use crate::vote::{DkgSignedVote, DkgVote, IdAck, IdPart, NodeId};
+
+/// State of the Dkg session, contains the sync keygen and currently known Parts and Acks
+/// Can handle votes coming from other participants
+pub struct DkgState<R: bls::rand::RngCore> {
+    id: NodeId,
+    secret_key: SecretKey,
+    pub_keys: BTreeMap<NodeId, PublicKey>,
+    keygen: SyncKeyGen<NodeId>,
+    our_part: Part,
+    all_votes: BTreeSet<DkgSignedVote>,
+    rng: R,
+}
+
+pub enum VoteResponse {
+    WaitingForMoreVotes,
+    BroadcastVote(Box<DkgSignedVote>),
+    RequestAntiEntropy,
+    AntiEntropy(BTreeSet<DkgSignedVote>),
+    DkgComplete(PublicKeySet, SecretKeyShare),
+}
+
+enum DkgCurrentState {
+    IncompatibleVotes,
+    NeedAntiEntropy,
+    Termination(BTreeMap<IdPart, BTreeSet<IdAck>>),
+    WaitingForTotalAgreement,
+    GotAllAcks(BTreeMap<IdPart, BTreeSet<IdAck>>),
+    WaitingForMoreAcks,
+    GotAllParts(BTreeSet<IdPart>),
+    WaitingForMoreParts,
+}
+
+impl<R: bls::rand::RngCore + Clone> DkgState<R> {
+    pub fn new(
+        our_id: NodeId,
+        secret_key: SecretKey,
+        pub_keys: BTreeMap<NodeId, PublicKey>,
+        threshold: usize,
+        rng: &mut R,
+    ) -> Result<Self> {
+        let (sync_key_gen, opt_part) = SyncKeyGen::new(
+            our_id,
+            secret_key.clone(),
+            Arc::new(pub_keys.clone()),
+            threshold,
+            rng,
+        )?;
+        Ok(DkgState {
+            id: our_id,
+            secret_key,
+            pub_keys,
+            keygen: sync_key_gen,
+            all_votes: BTreeSet::new(),
+            our_part: opt_part.ok_or(Error::NotInPubKeySet)?,
+            rng: rng.clone(),
+        })
+    }
+
+    /// The 1st vote with our Part
+    pub fn first_vote(&mut self) -> Result<DkgSignedVote> {
+        let vote = DkgVote::SinglePart(self.our_part.clone());
+        self.cast_vote(vote)
+    }
+
+    fn get_validated_vote(&self, vote: &DkgSignedVote) -> Result<DkgVote> {
+        let sender_id = vote.voter;
+        let sender_pub_key = self.pub_keys.get(&sender_id).ok_or(Error::UnknownSender)?;
+        let vote = vote.get_validated_vote(sender_pub_key)?;
+        Ok(vote)
+    }
+
+    fn all_checked_votes(&self) -> Result<Vec<(DkgVote, NodeId)>> {
+        self.all_votes
+            .iter()
+            .map(|v| Ok((self.get_validated_vote(v)?, v.voter)))
+            .collect()
+    }
+
+    fn current_dkg_state(&self, votes: Vec<(DkgVote, NodeId)>) -> DkgCurrentState {
+        let knowledge = match Knowledge::from_votes(votes) {
+            Err(KnowledgeFault::IncompatibleAcks) | Err(KnowledgeFault::IncompatibleParts) => {
+                return DkgCurrentState::IncompatibleVotes;
+            }
+            Err(KnowledgeFault::MissingParts) | Err(KnowledgeFault::MissingAcks) => {
+                return DkgCurrentState::NeedAntiEntropy;
+            }
+            Ok(k) => k,
+        };
+
+        let participants_len = self.pub_keys.len();
```

Comment:
Call this `num_participants` or `n_participants` to show it's the number of participants, rather than length